### PR TITLE
refactor detection, support detr and dab-detr

### PR DESCRIPTION
**Pull request by tuofeilunhifi**
1. refactor detection
2.support detr(39.92) and dab-detr(42.52)

### CODE REVIEW

File: easycv/apis/test.py [hunk 1]
```diff
@@ -211,7 +211,7 @@ def multi_gpu_test(model,
 
         if rank == 0:
             if 'img_metas' in data:
-                batch_size = len(data['img_metas'].data[0])
+                batch_size = len(data['img_metas'][0].data[0])
```
Comment (haiasd) [inline easycv/apis/test.py:214]: Modify within the algorithm？

File: configs/detection/vitdet/_base_/datasets/coco_instance.py [hunk 2]
```diff
@@ -15,7 +15,7 @@
 ]
 
 # dataset settings
-data_root = 'data/coco/'
+data_root = '/apsarapangu/disk2/yunji.cjy/coco/'
```
Comment (wenmengzhou) [inline configs/detection/vitdet/_base_/datasets/coco_instance.py:18]: reset to original value

File: configs/detection/yolox/yolox_s_8xb16_300e_coco.py [hunk 3]
```diff
@@ -35,7 +35,7 @@
 ]
 
 # dataset settings
-data_root = 'data/coco/'
+data_root = '/apsarapangu/disk2/yunji.cjy/coco/'
```
Comment (wenmengzhou) [inline configs/detection/yolox/yolox_s_8xb16_300e_coco.py:38]: revert to original value

File: easycv/models/detection/dab_detr/dab_detr_head.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
+# Copyright (c) 2022 IDEA. All Rights Reserved.
```
Comment (wenmengzhou) [inline easycv/models/detection/dab_detr/dab_detr_head.py:1]: add alibaba copyright

File: easycv/models/detection/dab_detr/dab_detr_transformer.py [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+            else:
+                return [
+                    torch.stack(intermediate).transpose(1, 2),
+                    reference_points.unsqueeze(0).transpose(1, 2)
+                ]
+
+        return output.unsqueeze(0)
+
+
+class TransformerEncoderLayer(nn.Module):
```
Comment (wenmengzhou) [inline easycv/models/detection/dab_detr/dab_detr_transformer.py:446]: we should reuse these blocks from detr if we could

File: easycv/models/detection/dab_detr/dab_detr_transformer.py [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+        src = src + self.dropout1(src2)
+        src = self.norm1(src)
+        src2 = self.linear2(self.dropout(self.activation(self.linear1(src))))
+        src = src + self.dropout2(src2)
+        src = self.norm2(src)
+        return src
+
+
+class TransformerDecoderLayer(nn.Module):
+
```
Comment (wenmengzhou) [inline easycv/models/detection/dab_detr/dab_detr_transformer.py:494]: same as above

File: easycv/models/detection/utils/boxes.py [hunk 7]
```diff
@@ -76,3 +78,332 @@ def postprocess(prediction, num_classes, conf_thre=0.7, nms_thre=0.45):
             output[i] = torch.cat((output[i], detections))
 
     return output
+
+
+def bbox2result(bboxes, labels, num_classes):
```
Comment (wenmengzhou) [inline easycv/models/detection/utils/boxes.py:83]: does these code borrowed from mmlab? we should add declaration

File: easycv/models/detection/utils/misc.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+        tuple(list): A tuple containing multiple list, each list contains \
+            a kind of returned results by the function
+    """
+    pfunc = partial(func, **kwargs) if kwargs else func
+    map_results = map(pfunc, *args)
+    return tuple(map(list, zip(*map_results)))
+
+
+@torch.no_grad()
+def accuracy(output, target, topk=(1, )):
```
Comment (wenmengzhou) [inline easycv/models/detection/utils/misc.py:37]: where should these be used, if we need accuracy, we should add it to evaluators

File: easycv/apis/test.py (hunk 1 above)
Comment (tuofeilunhifi) [reply easycv/apis/test.py]: done

File: easycv/datasets/detection/data_sources/coco.py [hunk 9]
```diff
@@ -68,10 +68,8 @@ def get_length(self):
 
     def load_annotations(self, ann_file):
         """Load annotation from COCO style annotation file.
-
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/datasets/detection/data_sources/coco.py:71]: done

File: configs/detection/dab_detr/dab_detr.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# model settings
+model = dict(
+    type='Detection',
+    pretrained=
+    'http://pai-vision-data-hz.oss-cn-zhangjiakou.aliyuncs.com/pretrained_models/easycv/resnet/torchvision/resnet50.pth',
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply configs/detection/dab_detr/dab_detr.py]: done

File: configs/detection/vitdet/_base_/datasets/coco_instance.py (hunk 2 above)
Comment (tuofeilunhifi) [reply configs/detection/vitdet/_base_/datasets/coco_instance.py:18]: done

File: configs/detection/yolox/yolox_s_8xb16_300e_coco.py (hunk 3 above)
Comment (tuofeilunhifi) [reply configs/detection/yolox/yolox_s_8xb16_300e_coco.py:38]: done

File: easycv/models/builder.py [hunk 11]
```diff
@@ -37,3 +38,7 @@ def build_loss(cfg):
 
 def build_model(cfg):
     return build(cfg, MODELS)
+
+
+def build_transformer(cfg):
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/models/builder.py]: done

File: easycv/models/detection/dab_detr/dab_detr_head.py (hunk 4 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/dab_detr/dab_detr_head.py:1]: done

File: easycv/models/detection/dab_detr/dab_detr_transformer.py (hunk 5 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/dab_detr/dab_detr_transformer.py]: done

File: easycv/models/detection/dab_detr/dab_detr_transformer.py (hunk 6 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/dab_detr/dab_detr_transformer.py:422]: done

File: easycv/models/detection/utils/boxes.py (hunk 7 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/utils/boxes.py:84]: done

File: easycv/models/detection/utils/misc.py (hunk 8 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/utils/misc.py:37]: detr series debug is used to print logs

Comment (tuofeilunhifi) [reply easycv/models/detection/utils/misc.py:37]: Used when the DETR series prints logs

File: easycv/models/detection/dab_detr/dab_detr_head.py (hunk 4 above)
Comment (wenmengzhou) [reply easycv/models/detection/dab_detr/dab_detr_head.py:1]: IDEA and Alibaba

File: easycv/models/detection/utils/__init__.py [hunk 12]
```diff
@@ -1,5 +1,12 @@
 #!/usr/bin/env python3
 # Copyright (c) 2014-2021 Megvii Inc. All rights reserved.
-
-from .boxes import bboxes_iou, postprocess
-from .utils import output_postprocess
+from .boxes import (bbox2result, bbox_overlaps, bboxes_iou, box_cxcywh_to_xyxy,
```
Comment (Cathy0908) [inline easycv/models/detection/utils/__init__.py:3]: Check that all common apis under utils directory, and  move to common utils

File: easycv/utils/fp16_utils.py [hunk 13]
```diff
@@ -0,0 +1,121 @@
+# Copyright (c) OpenMMLab. All rights reserved.
```
Comment (Cathy0908) [inline easycv/utils/fp16_utils.py:1]: remove this script

File: easycv/datasets/detection/data_sources/coco.py [hunk 14]
```diff
@@ -71,7 +71,6 @@ def load_annotations(self, ann_file):
 
         Args:
             ann_file (str): Path of annotation file.
-
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/datasets/detection/data_sources/coco.py]: pre-commit auto remove

File: easycv/core/evaluation/coco_tools.py [hunk 15]
```diff
@@ -546,7 +546,6 @@ def ExportSingleImageDetectionBoxesToCoco(image_id, category_id_set,
       do not have the right lengths or (2) if each of the elements inside these
       lists do not have the correct shapes or (3) if image_ids are not integers.
   """
-
```
[the remark this answers is not in the document]
Comment (tuofeilunhifi) [reply easycv/core/evaluation/coco_tools.py]: pre-commit auto remove

File: easycv/utils/fp16_utils.py (hunk 13 above)
Comment (tuofeilunhifi) [reply easycv/utils/fp16_utils.py]: done

File: easycv/models/detection/utils/__init__.py (hunk 12 above)
Comment (tuofeilunhifi) [reply easycv/models/detection/utils/__init__.py:3]: done

**Review by wenmengzhou — APPROVED** [review]
